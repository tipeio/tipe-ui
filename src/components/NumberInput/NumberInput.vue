<template>
  <div
    :data-tipe-ui="$options.name"
    :class="rootClassObject"
  >
    <input
      ref="input"
      :placeholder="placeholder"
      :value="displayValue"
      :disabled="disabled"
      :tabindex="tabindex"
      :max="max"
      :min="min"
      :step="step"
      type="number"
      @focus="onFocus"
      @blur="onBlur"
      @click="$emit('click', $event)"
      @change="onChange"
      @input="onInput"
    >
    <button
      :disabled="disabled"
      @mousedown.stop.prevent
      @click="onIncrement"
    >
      <tipe-icon icon="arrow-up" />
    </button>
    <button
      :disabled="disabled"
      @mousedown.stop.prevent
      @click="onDecrement"
    >
      <tipe-icon icon="arrow-down"/>
    </button>
  </div>
</template>

<script>
import vueTypes from 'vue-types'
import interfaces from '@/interfaces'
import TipeIcon from '@/components/Icon'
import { getPrecision } from '@/libs/float'

export default {
  name: 'TipeNumberInput',
  components: {
    TipeIcon
  },
  props: {
    value: vueTypes.oneOfType([vueTypes.number, undefined]),
    max: vueTypes.number.def(Infinity),
    min: vueTypes.number.def(-Infinity),
    step: vueTypes.number.def(1),
    placeholder: vueTypes.oneOfType([vueTypes.number, vueTypes.string]),
    ...interfaces.input
  },
  data() {
    return {
      displayValue: this.value,
      focus: false
    }
  },
  computed: {
    rootClassObject() {
      return {
        [`status-${this.status}`]: this.status,
        waiting: this.waiting,
        [this.size]: true,
        focus: this.focus
      }
    }
  },
  methods: {
    getDisplayValue() {
      const displayValue = parseFloat(this.displayValue)
      return isNaN(displayValue) ? undefined : displayValue
    },
    setDisplayValue(value) {
      this.displayValue = Math.min(
        Math.max(
          parseFloat(value).toFixed(
            Math.min(getPrecision(this.step), getPrecision(value))
          ),
          this.min
        ),
        this.max
      )
    },
    onInput(event) {
      this.setDisplayValue(event.target.value)
      this.$emit('input', this.displayValue)
    },
    onChange(event) {
      this.setDisplayValue(event.target.value)
      this.$emit('change', this.displayValue)
    },
    onFocus(event) {
      this.focus = true
      this.$emit('focus', event)
    },
    onBlur(event) {
      this.focus = false
      this.$emit('blur', event)
    },
    onIncrement() {
      this.$refs.input.focus()
      const value = this.getDisplayValue()

      if (value === undefined) {
        if (this.min === -Infinity) {
          this.setDisplayValue(0)
        } else {
          this.setDisplayValue(this.min)
        }
      } else if (value + this.step <= this.max) {
        this.setDisplayValue(value + this.step)
      } else {
        return
      }

      this.$emit('change', this.displayValue)
    },
    onDecrement() {
      this.$refs.input.focus()
      const value = this.getDisplayValue()

      if (value === undefined) {
        if (this.min === -Infinity) {
          this.setDisplayValue(0)
        } else {
          this.setDisplayValue(this.min)
        }
      } else if (value - this.step >= this.min) {
        this.setDisplayValue(value - this.step)
      } else {
        return
      }

      this.$emit('change', this.displayValue)
    }
  }
}
</script>

<style lang="postcss" scoped>
[data-tipe-ui='TipeNumberInput'] {
  display: grid;
  grid-template-areas:
    'input up'
    'input down';
  grid-template-columns: 1fr 2.4375rem;
  grid-template-rows: 1fr 1fr;
  border: 1px solid var(--gray-light);
  border-radius: 0.25rem;
  overflow: hidden;
  width: 10rem;
}

[data-tipe-ui='TipeNumberInput'].small {
  line-height: 0.6875rem;
  font-size: 0.6875rem;
  height: 1.625rem;
}

[data-tipe-ui='TipeNumberInput'].medium {
  line-height: 0.875rem;
  font-size: 0.8125rem;
  height: 2.5rem;
}

[data-tipe-ui='TipeNumberInput'].large {
  font-size: 1rem;
  height: 3rem;
}

[data-tipe-ui='TipeNumberInput'].status-error {
  border: 1px solid var(--error);
}

[data-tipe-ui='TipeNumberInput'].status-warning {
  border: 1px solid var(--warning);
}

[data-tipe-ui='TipeNumberInput'].status-success {
  border: 1px solid var(--success);
}

[data-tipe-ui='TipeNumberInput'].focus {
  border: 1px solid var(--gray);
}

[data-tipe-ui='TipeNumberInput'].focus.status-error {
  border: 1px solid var(--danger-active);
}

[data-tipe-ui='TipeNumberInput'].focus.status-warning {
  border: 1px solid var(--warning-active);
}

[data-tipe-ui='TipeNumberInput'].focus.status-success {
  border: 1px solid var(--success-active);
}

input:disabled,
button:disabled {
  cursor: not-allowed;
  opacity: 0.6;
}

input {
  grid-area: input;
  -moz-appearance: textfield;
  outline: none;
  background-color: #fff;
  margin: 0;
  padding: 0;
  height: 100%;
  width: 100%;
  font-size: inherit;
  letter-spacing: 0.00375rem;
  border: none;
  border-top-left-radius: 0.25rem;
  border-bottom-left-radius: 0.25rem;
  box-sizing: border-box;
  padding: 0.1875rem 0.8125rem;
  text-align: center;
  color: var(--dark-purple);
}

input[type='number']::-webkit-inner-spin-button,
input[type='number']::-webkit-outer-spin-button {
  -webkit-appearance: none;
  margin: 0;
}

button {
  display: flex;
  justify-content: center;
  align-items: center;
  height: 100%;
  width: 100%;
  margin: 0;
  padding: 0;
  outline: none;
  background-color: #f7f8fb;
  box-sizing: border-box;
  border: none;
  border-left: 1px solid var(--gray-light);
}

button:first-of-type {
  grid-area: up;
  border-bottom: 1px solid var(--gray-light);
  border-top-right-radius: 0.25rem;
}

button:last-of-type {
  grid-area: down;
  border-bottom-right-radius: 0.25rem;
}
</style>
