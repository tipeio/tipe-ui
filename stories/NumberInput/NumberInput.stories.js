import { storiesOf } from '@storybook/vue'
import { withKnobs, select, number, boolean } from '@storybook/addon-knobs/vue'
import { action } from '@storybook/addon-actions'
import { withMarkdownNotes } from '@storybook/addon-notes'

import TipeNumberInput from '@/components/NumberInput'
import { text } from '../../node_modules/@storybook/addon-knobs/dist/vue'

const notes = `
#### :props
:value | number |
:select | string | mini, small, medium, large
:min | number |
:max | number |
:step | number |

#### @events
@click($event)
@focus($event)
@blur($event)
@change($event)
`

storiesOf('NumberInput', module)
  .addDecorator(withKnobs)
  .add(
    'default',
    withMarkdownNotes(notes)(() => ({
      components: { TipeNumberInput },
      data() {
        return {
          value: number('value', 3),
          placeholder: text('placeholder', 'Quantity'),
          size: select('size', ['mini', 'small', 'medium', 'large'], 'medium'),
          min: number('min', 0),
          max: number('max', 100),
          step: number('step', 1),
          status: select('status', ['', 'success', 'warning', 'error'], ''),
          disabled: boolean('disabled', false),
          waiting: boolean('waiting', false)
        }
      },
      methods: {
        click: action('click'),
        focus: action('focus'),
        blur: action('blur'),
        change: action('change')
      },
      template: `
        <tipe-number-input
          :value="value"
          :placeholder="placeholder"
          :size="size"
          :min="min"
          :max="max"
          :step="step"
          :status="status"
          :disabled="disabled"
          :waiting="waiting"
          @click="click"
          @focus="focus"
          @blur="blur"
          @change="change"
        />`
    }))
  )
