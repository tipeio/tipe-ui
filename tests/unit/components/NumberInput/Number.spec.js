import { shallowMount } from '@vue/test-utils'
import { createRenderer } from 'vue-server-renderer'
import NumberInput from '@/components/NumberInput'
import tipeTestUtils from '~/tests/TipeTestUtils'
import faker from 'faker'

describe('NumberInput', () => {
  describe('<template>', () => {
    it('matches previous snapshot', () => {
      const renderer = createRenderer()
      const propsData = { file: { type: 'document', label: 'test' } }
      const wrapper = shallowMount(NumberInput, { propsData })
      renderer.renderToString(wrapper.vm, (err, str) => {
        if (err) throw new Error(err)
        expect(str).toMatchSnapshot()
      })
    })

    it('has correct data-tipe-iu attibute', () => {
      const wrapper = shallowMount(NumberInput)

      expect(wrapper.attributes()['data-tipe-ui']).toBe('TipeNumberInput')
    })

    it('contains an input tag', () => {
      const wrapper = shallowMount(NumberInput)

      expect(wrapper.contains('input')).toEqual(true)
    })

    it('input type is number', () => {
      const wrapper = shallowMount(NumberInput)

      expect(wrapper.find('input').attributes().type).toEqual('number')
    })

    it('contains spinner buttons', () => {
      const wrapper = shallowMount(NumberInput)

      expect(wrapper.findAll('button')).toHaveLength(2)
    })
  })
  describe(':props', () => {
    it(':value - should pass value input element', () => {
      const propsData = { value: faker.random.number() }
      const wrapper = shallowMount(NumberInput, { propsData })
      expect(wrapper.find('input').element.value).toEqual(`${propsData.value}`)
    })

    tipeTestUtils.test.input.size(NumberInput)

    tipeTestUtils.test.input.status(NumberInput)

    tipeTestUtils.test.input.disabled(NumberInput, wrapper =>
      wrapper.find('input')
    )

    it(':placeholder - should pass placeholder to input', () => {
      const propsData = { placeholder: 'number here' }
      const wrapper = shallowMount(NumberInput, { propsData })
      expect(wrapper.find('input').attributes().placeholder).toEqual(
        `${propsData.placeholder}`
      )
    })

    tipeTestUtils.test.input.waiting(NumberInput)

    it(':waiting(true) - has disabled attribute', () => {
      const propsData = { disabled: true }
      const wrapper = shallowMount(NumberInput, { propsData })
      expect(wrapper.find('input').attributes().disabled).toBeTruthy()
      expect(wrapper.find('button').attributes().disabled).toBeTruthy()
      expect(
        wrapper
          .findAll('button')
          .at(1)
          .attributes().disabled
      ).toBeTruthy()
    })

    it(':step - passed to input step attribue', () => {
      const propsData = { step: faker.random.number() }
      const wrapper = shallowMount(NumberInput, { propsData })
      expect(wrapper.find('input').attributes().step).toEqual(
        `${propsData.step}`
      )
    })

    it(':max - passed to input max attribue', () => {
      const propsData = { max: faker.random.number() }
      const wrapper = shallowMount(NumberInput, { propsData })
      expect(wrapper.find('input').attributes().max).toEqual(`${propsData.max}`)
    })

    it(':min - passed to input max attribue', () => {
      const propsData = { min: faker.random.number() }
      const wrapper = shallowMount(NumberInput, { propsData })
      expect(wrapper.find('input').attributes().min).toEqual(`${propsData.min}`)
    })
  })

  describe('@events', () => {
    it('@focus - should emit', () => {
      const wrapper = shallowMount(NumberInput)
      wrapper.find('input').trigger('focus')
      expect(wrapper.emitted().focus).toBeTruthy()
    })

    it('@blur - should emit', () => {
      const wrapper = shallowMount(NumberInput)
      wrapper.find('input').trigger('blur')
      expect(wrapper.emitted().blur).toBeTruthy()
    })

    it('@click - should emit', () => {
      const wrapper = shallowMount(NumberInput)
      wrapper.find('input').trigger('click')
      expect(wrapper.emitted().click).toBeTruthy()
    })

    it('@change - should emit', () => {
      const wrapper = shallowMount(NumberInput)
      wrapper.find('input').trigger('change')
      expect(wrapper.emitted().change).toBeTruthy()
    })

    it('@input - should emit', () => {
      const wrapper = shallowMount(NumberInput)
      wrapper.find('input').trigger('input')
      expect(wrapper.emitted().input).toBeTruthy()
    })

    it('<button @click> - first button increments value', () => {
      const wrapper = shallowMount(NumberInput, { propsData: { value: 0 } })
      const input = wrapper.find('input')
      const originalValue = parseInt(input.element.value)

      wrapper.find('button').trigger('click')

      expect(parseInt(input.element.value)).toEqual(originalValue + 1)
    })

    it('<button @click> - second button decrements value', () => {
      const wrapper = shallowMount(NumberInput, { propsData: { value: 0 } })
      const input = wrapper.find('input')
      const originalValue = parseInt(input.element.value)

      wrapper
        .findAll('button')
        .at(1)
        .trigger('click')

      expect(parseInt(input.element.value)).toEqual(originalValue - 1)
    })
  })
})
