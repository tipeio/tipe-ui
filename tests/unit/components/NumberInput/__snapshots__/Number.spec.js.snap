// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NumberInput <template> matches previous snapshot 1`] = `
<div data-tipe-ui="TipeNumberInput" class="medium">
  <input tabindex="0" max="Infinity" min="-Infinity" step="1" type="number">
  <button>
    <TipeIcon-stub></TipeIcon-stub>
  </button>
  <button>
    <TipeIcon-stub></TipeIcon-stub>
  </button>
</div>
`;
